### ISSUE TITLE
CEF Upgrade

### ISSUE THREAD

**Issue by Julusian** [CasparCG/server#1235]
### Description
CEF often has breaking bugs which block us from upgrading to certain versions. The purpose of this issue is to track those bugs so that we know why we can't use a version and what to verify is fixed when we do upgrade.

#### CEF 73
Nothing known

#### CEF 75
* HTML video doesnt trigger repaints, causing some pages (such as youtube) to not render smoothly/at all. (Might affect 74 too)

#### CEF 76
* [New Viz compositor](https://bitbucket.org/chromiumembedded/cef/issues/2575/viz-implementation-for-osr). Removes shared_texture functionality. [Issue](https://bitbucket.org/chromiumembedded/cef/issues/1006/cef3-support-rendering-to-a-hardware-gl)

**Comment by Jgaleas7** [comment]
in the version 77??

**Comment by sirfnomi** [comment]
I think shared texture still not back in 77

**Comment by sirfnomi** [comment]
Hi Julusian, Is CasparCG updating with 73 till new Viz gets shared texture ?

**Comment by Julusian** [comment]
Yes, my intention is for #1217 to update to CEF73. Hopefully that will happen sometime this month

**Comment by sirfnomi** [comment]
Thanks hopefully.

**Comment by Jgaleas7** [comment]
any update?

**Comment by Julusian** [comment]
@Jgaleas7 master is now updated to CEF71, with shared texture support

**Comment by dotarmin** [comment]
> @Jgaleas7 master is now updated to CEF71, with shared texture support

I'm sure the builds has stoped to work due to this. I can try to fix it this weekend or on Monday morning.

**Comment by grahamspr86** [comment]
> > @Julusian  master is now updated to CEF71, with shared texture support
> 
> I'm sure the builds has stoped to work due to this. I can try to fix it this weekend or on Monday morning.

That explains why I couldn't see a build! Wasn't sure how long they took following the commit

**Comment by dotarmin** [comment]
> That explains why I couldn't see a build! Wasn't sure how long they took following the commit

I just noticed it, I can maybe even try to fix it tonight if I'm free. It takes around 7 minutes for a build to finish.

**Comment by Julusian** [comment]
@dotarmin It compiles in vs fine, so I guess it will be something with the packaging? The files needed to be copied from CEF has changed in this version so that could be it, but its hard to say without seeing the build output

**Comment by dotarmin** [comment]
> The files needed to be copied from CEF has changed in this version so that could be it

Exactly what it is. I have an idea about having a repo with build-scripts that checks out and build the server OR we put back the build scripts into this repo to make it "self living", then on checkout we can use the scripts available within the repo it-self.

Input?

**Comment by Julusian** [comment]
I think that a script which simply creates a built zip in this repo would make sense. Especially as the script now will need to be a little different for 2.2 and 2.3. Versioning of it will get complicated if it lives elsewhere.

**Comment by dotarmin** [comment]
> Under `tools/windows/` is a good place for it, as we already have linux ones next to that

Great @Julusian, I can try fix this later or this weekend. Otherwise I can do it next week. The scripts will be similar to the once we had before in v2.1.x, slightly modified. 

Thanks!

**Comment by krushd** [comment]
@Julusian @dotarmin - You may have already seen this, but looks like OBS finally got a newer CEF package to work with shared texture support - maybe a path forward for updating CasparCG: https://github.com/obsproject/obs-studio/discussions/3853#discussioncomment-1710556

**Comment by mryvz** [comment]
@krushd @Julusian @dotarmin - I've migrated to CEF version 97 master branch and looks like its working but i cannot be sure its ok; 
* play 1-1 [html] get.webgl.org displays and says your browser supports webgl
* no gpu memory leak on repeating play html template commands 
* pixijs examples plays well 

but;
* onacceleratedpaint never invoked, only onpaint  function invoked as if gpu enabled false in config. (i ve disabled shared_texture_enabled check block )
* there is a console log says "0220/014656.393:ERROR:gpu_init.cc(457)] Passthrough is not supported, GL is disabled, ANGLE"

**Comment by hreinnbeck** [comment]
What changes did you do to the HTML module?

Every CEF upgrade has involved refactoring of the HTML module, this big of a jump in version might even require significant refactoring. But it sounds promising that you got it working.

You should get some hints by checking what modifications obs-browser did when they upgraded CEF.  If you are running on Win you might want to see if adding a win10 manifest helps (https://github.com/nrkno/sofie-casparcg-server/pull/41/commits/4f432eb69c794c070af0828c05349cd6070bdeba)

**Comment by mryvz** [comment]
@hreinnbeck  Not Much only basic  things, i will push changes later , for now changes below, 
`
html_producer.cpp
* OnBeforePopup method has an extra argument so just changed overrided mehod to new form,
* OnProcessMessageReceived method has an extra  argument so just changed override method to new form.
*  CefBrowser::SendProcessMessage method no more exist replaced all to  CefBrowser::GetMainFrame::SendProcessMessage 
* CefWindowInfo.width and CefWindowInfo.height no more , replaced to CefWindowInfo.bounds.width and CefWindowInfo.bounds.height
* CefBrowserHost::CreateBrowser has an extra argument , just passed nullptr
`
html.cpp
* OnProcessMessageReceived method has an extra  argument so just changed override method to new form.
*  CefBrowser::SendProcessMessage method no more exist replaced all to  CefBrowser::GetMainFrame::SendProcessMessage

**Comment by Julusian** [comment]
I dont know how I missed these comments..

@mryvz thats because shared_texture support is still broken in CEF (since 76). There is an unmerged PR waiting since soon after it broke to reimplement it.

I have been playing with upgrading to 95, using the custom builds that OBS use, and it looks to be working well

**Comment by pbelbin** [comment]
Any word on upgrading to latest CEF?  It's kinda painful in some cases to not be able to use more recent javascript standards like module imports.

**Comment by sirfnomi** [comment]
I'm waiting for it too, hope it gets updated soon

**Comment by Julusian** [comment]
#1427

**Comment by Julusian** [comment]
master is now on CEF v95

OBS are putting a considerable effort into keeping their version of CEF updated (they jumped to 103 between me opening #1427 and merging it), so hopefully we can keep reusing their work for future updates

**Comment by krushd** [comment]
@Julusian I'm curious if the requestor of the OnAcceleratedPaint2 change is using it for a specific edge case that wouldn't impact Caspar (or OBS for that matter)? Tracing the thread back to https://github.com/obsproject/cef/commit/1cc2af87cf6bbdac9e1dbccb5efdfd1934c63062 - it seems like a new request, not something that was in and then removed in 5060 (which sounds like it will be the last version for a little while)). I got 103 to compile into Caspar - running some limited tests now.

**Comment by Julusian** [comment]
@krushd it may well work, all I tried was updating CEF, got no frames and gave up https://github.com/CasparCG/server/commit/ad5a317bf62b3f551cd43058d035243d09d9e05d
95 is new enough for my client, so I couldnt justify spending much time to try and get 103 working

**Comment by krushd** [comment]
@Julusian Same here (95 does everything I need) - got 103 to compile - but not getting anything real to play out other than chrome;//version.

I'll play around a little bit and post any updates here if I find anything,

**Comment by Julusian** [comment]
@krushd Upon getting black, my assumption (I may have set breakpoints too?) was that it wasnt calling `OnAcceleratedPaint2`. Presumably because it was reusing the d3d texture, which matches up perfectly with that linked PR.

For OBS, that doesnt matter as they directly pass the texture handle to their compositor. But we are buffering frames, and do a copy of the texture onto our own opengl texture as part of `OnAcceleratedPaint2`. So if its not called, we don't get a new frame.

**Comment by Sidonai-1** [comment]
I usually only download the builds on https://github.com/CasparCG/server/releases

Any good reason to start using more the builds on https://builds.casparcg.com/ to get the newer CEF 95 vs. 71? Or is it still too risky?

I assume the github releases page is going to be updated with CEF 95 and everything else once it gets some more testing, correct me if I'm wrong.

Thanks!
